panic: runtime error: invalid memory address or nil pointer dereference

Hello. Running latest nxs-backup 3.8.0. When I try validate my cnf  I get the following errors:

```
# nxs-backup --version
nxs-backup 3.8.0
```

```
# nxs-backup -t
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0xd59863]

goroutine 1 [running]:
github.com/nixys/nxs-backup/modules/metrics.(*Data).GetMetrics(...)
        /home/runner/work/nxs-backup/nxs-backup/modules/metrics/metrics.go:81
github.com/nixys/nxs-backup/modules/backup/psql.Init({{0xc000477350, 0xe}, {0xc000036ed0, 0x18}, 0x1, 0x0, 0x0, 0x0, {0xc00033fd50, 0x1, ...}, ...})
        /home/runner/work/nxs-backup/nxs-backup/modules/backup/psql/psql.go:97 +0x2a3
github.com/nixys/nxs-backup/ctx.jobsInit({0x0, 0x0, 0x0, {0xc00020eab0, 0x1, 0x1}, 0xc00035f200})
        /home/runner/work/nxs-backup/nxs-backup/ctx/jobs_init.go:276 +0x22f8
github.com/nixys/nxs-backup/ctx.appInit(0x3?, {0x103a1fb, 0x1f})
        /home/runner/work/nxs-backup/nxs-backup/ctx/context.go:210 +0x65a
github.com/nixys/nxs-backup/ctx.AppCtxInit()
        /home/runner/work/nxs-backup/nxs-backup/ctx/context.go:96 +0xa93
github.com/nixys/nxs-go-appctx/v3.(*AppCtx).Run(0xc0004e8d20)
        /home/runner/go/pkg/mod/github.com/nixys/nxs-go-appctx/v3@v3.0.0/appctx.go:65 +0x42
main.main()
        /home/runner/work/nxs-backup/nxs-backup/main.go:38 +0x4f5
```

My nxs-backup.conf: 

```
project_name: PROJECT
server_name: postgresql

server:
  metrics:
    enabled: false

loglevel: debug

notifications:
  mail:
    enabled: false
  webhooks:
    - webhook_url: "SOME URL"
      message_level: "warning"
      extra_headers:
        "X-Auth-Key": "SOME KEY"
      payload_message_key: "triggerMessage"
      extra_payload:
        "isEmergencyAlert": false
        "rawTriggerMessage": false
        "monitoringURL": "-"

include_jobs_configs: ["conf.d/*.conf"]
```

When I commented out the metrics: 

```
#server:
#  metrics:
#    enabled: false
```

I received the following error: 

```
# nxs-backup -t
Failed to init metrics: open /tmp/nxs-backup.metrics: permission denied
```

Fixed by: 

```
chown root: /tmp/nxs-backup.metrics
```

```
# nxs-backup -t
The configuration is correct

No external jobs
List of databases jobs:
  postgresql
No files jobs
```

--- a/ctx/context.go
+++ b/ctx/context.go
@@ -156,7 +156,6 @@ func printInitError(ft string, err error) {
 }
 
 func appInit(c *Ctx, cfgPath string) (app, error) {
-	var oldMetrics *metrics.Data
 
 	a := app{
 		jobs: make(map[string]interfaces.Job),
@@ -171,6 +170,15 @@ func appInit(c *Ctx, cfgPath string) (app, error) {
 	a.waitTimeout = conf.WaitingTimeout
 	a.serverBind = conf.Server.Bind
 
+	a.metricsData = metrics.InitData(
+		metrics.DataOpts{
+			Project:     conf.ProjectName,
+			Server:      conf.ServerName,
+			MetricsFile: conf.Server.Metrics.FilePath,
+			Enabled:     false,
+		},
+	)
+
 	if conf.Server.Metrics.Enabled {
 		nva := 0.0
 		ver, _ := semver.NewVersion(misc.VERSION)
@@ -178,18 +186,8 @@ func appInit(c *Ctx, cfgPath string) (app, error) {
 		if newVer != "" {
 			nva = 1
 		}
-		a.metricsData, oldMetrics, err = metrics.InitData(
-			metrics.DataOpts{
-				Project:             conf.ProjectName,
-				Server:              conf.ServerName,
-				MetricsFile:         conf.Server.Metrics.FilePath,
-				NewVersionAvailable: nva,
-			},
-		)
-		if err != nil {
-			printInitError("Failed to init metrics: %v\n", err)
-			return a, err
-		}
+		a.metricsData.NewVersionAvailable = nva
+		a.metricsData.Enabled = true
 	}
 
 	if err = logInit(c, conf.LogFile, conf.LogLevel); err != nil {
@@ -209,11 +207,10 @@ func appInit(c *Ctx, cfgPath string) (app, error) {
 
 	jobs, err := jobsInit(
 		jobsOpts{
-			jobs:           conf.Jobs,
-			storages:       storages,
-			metricsData:    a.metricsData,
-			oldMetricsData: oldMetrics,
-			mainLim:        conf.Limits,
+			jobs:        conf.Jobs,
+			storages:    storages,
+			metricsData: a.metricsData,
+			mainLim:     conf.Limits,
 		},
 	)
 	if err != nil {

--- a/ctx/jobs_init.go
+++ b/ctx/jobs_init.go
@@ -27,11 +27,10 @@ import (
 )
 
 type jobsOpts struct {
-	metricsData    *metrics.Data
-	oldMetricsData *metrics.Data
-	mainLim        *limitsConf
-	jobs           []jobConf
-	storages       map[string]interfaces.Storage
+	metricsData *metrics.Data
+	mainLim     *limitsConf
+	jobs        []jobConf
+	storages    map[string]interfaces.Storage
 }
 
 func jobsInit(o jobsOpts) ([]interfaces.Job, error) {
@@ -135,7 +134,6 @@ func jobsInit(o jobsOpts) ([]interfaces.Job, error) {
 				Storages:         jobStorages,
 				Sources:          sources,
 				Metrics:          o.metricsData,
-				OldMetrics:       o.oldMetricsData,
 			})
 
 		case misc.IncFiles:
@@ -160,7 +158,6 @@ func jobsInit(o jobsOpts) ([]interfaces.Job, error) {
 				Storages:        jobStorages,
 				Sources:         sources,
 				Metrics:         o.metricsData,
-				OldMetrics:      o.oldMetricsData,
 			})
 
 		case misc.Mysql:
@@ -200,7 +197,6 @@ func jobsInit(o jobsOpts) ([]interfaces.Job, error) {
 				Storages:         jobStorages,
 				Sources:          sources,
 				Metrics:          o.metricsData,
-				OldMetrics:       o.oldMetricsData,
 			})
 
 		case misc.MysqlXtrabackup:
@@ -241,7 +237,6 @@ func jobsInit(o jobsOpts) ([]interfaces.Job, error) {
 				Storages:         jobStorages,
 				Sources:          sources,
 				Metrics:          o.metricsData,
-				OldMetrics:       o.oldMetricsData,
 			})
 
 		case misc.Postgresql:
@@ -283,7 +278,6 @@ func jobsInit(o jobsOpts) ([]interfaces.Job, error) {
 				Storages:         jobStorages,
 				Sources:          sources,
 				Metrics:          o.metricsData,
-				OldMetrics:       o.oldMetricsData,
 			})
 
 		case misc.PostgresqlBasebackup:
@@ -323,7 +317,6 @@ func jobsInit(o jobsOpts) ([]interfaces.Job, error) {
 				Storages:         jobStorages,
 				Sources:          sources,
 				Metrics:          o.metricsData,
-				OldMetrics:       o.oldMetricsData,
 			})
 
 		case misc.MongoDB:
@@ -366,7 +359,6 @@ func jobsInit(o jobsOpts) ([]interfaces.Job, error) {
 				Storages:         jobStorages,
 				Sources:          sources,
 				Metrics:          o.metricsData,
-				OldMetrics:       o.oldMetricsData,
 			})
 
 		case misc.Redis:
@@ -395,7 +387,6 @@ func jobsInit(o jobsOpts) ([]interfaces.Job, error) {
 				Storages:         jobStorages,
 				Sources:          sources,
 				Metrics:          o.metricsData,
-				OldMetrics:       o.oldMetricsData,
 			})
 
 		case misc.External:
@@ -407,7 +398,6 @@ func jobsInit(o jobsOpts) ([]interfaces.Job, error) {
 				SkipBackupRotate: j.SkipBackupRotate,
 				Storages:         jobStorages,
 				Metrics:          o.metricsData,
-				OldMetrics:       o.oldMetricsData,
 			})
 
 		default:

--- a/interfaces/job.go
+++ b/interfaces/job.go
@@ -7,7 +7,6 @@ import (
 
 type Job interface {
 	SetOfsMetrics(ofs string, metrics map[string]float64)
-	ExportMetrics()
 	GetName() string
 	GetTempDir() string
 	GetType() misc.BackupType

--- a/modules/backup/backup.go
+++ b/modules/backup/backup.go
@@ -49,8 +49,6 @@ func Perform(logCh chan logger.LogRecord, job interfaces.Job) error {
 		errs = multierror.Append(errs, err)
 	}
 
-	job.ExportMetrics()
-
 	_ = job.CleanupTmpData()
 	_ = filepath.Walk(tmpDirPath,
 		func(path string, info os.FileInfo, err error) error {

--- a/modules/backup/desc_files/desc_files.go
+++ b/modules/backup/desc_files/desc_files.go
@@ -31,7 +31,6 @@ type job struct {
 	targets          map[string]target
 	dumpedObjects    map[string]interfaces.DumpObject
 	appMetrics       *metrics.Data
-	jobMetrics       metrics.JobData
 }
 
 type target struct {
@@ -51,7 +50,6 @@ type JobParams struct {
 	Storages         interfaces.Storages
 	Sources          []SourceParams
 	Metrics          *metrics.Data
-	OldMetrics       *metrics.Data
 }
 
 type SourceParams struct {
@@ -79,15 +77,14 @@ func Init(jp JobParams) (interfaces.Job, error) {
 		storages:         jp.Storages,
 		targets:          make(map[string]target),
 		dumpedObjects:    make(map[string]interfaces.DumpObject),
-		appMetrics:       jp.Metrics,
-		jobMetrics: metrics.JobData{
-			JobName:       jp.Name,
-			JobType:       misc.DescFiles,
-			TargetMetrics: make(map[string]metrics.TargetData),
-		},
+		appMetrics: jp.Metrics.RegisterJob(
+			metrics.JobData{
+				JobName:       jp.Name,
+				JobType:       misc.DescFiles,
+				TargetMetrics: make(map[string]metrics.TargetData),
+			},
+		),
 	}
-
-	ojm := jp.OldMetrics.GetMetrics(jp.Name)
 
 	for _, src := range jp.Sources {
 
@@ -128,32 +125,23 @@ func Init(jp JobParams) (interfaces.Job, error) {
 						saveAbsPath: src.SaveAbsPath,
 						excludes:    excludes,
 					}
-					if otm, ok := ojm.TargetMetrics[ofs]; ok {
-						j.jobMetrics.TargetMetrics[ofs] = otm
-					} else {
-						j.jobMetrics.TargetMetrics[ofs] = metrics.TargetData{
-							Source: src.Name,
-							Target: ofsPart,
-							Values: make(map[string]float64),
-						}
+					j.appMetrics.Job[jp.Name].TargetMetrics[ofs] = metrics.TargetData{
+						Source: src.Name,
+						Target: ofsPart,
+						Values: make(map[string]float64),
 					}
 				}
 			}
 		}
 	}
 
-	j.ExportMetrics()
 	return &j, nil
 }
 
 func (j *job) SetOfsMetrics(ofs string, metricsMap map[string]float64) {
 	for m, v := range metricsMap {
-		j.jobMetrics.TargetMetrics[ofs].Values[m] = v
+		j.appMetrics.Job[j.name].TargetMetrics[ofs].Values[m] = v
 	}
-}
-
-func (j *job) ExportMetrics() {
-	j.appMetrics.JobMetricsSet(j.jobMetrics)
 }
 
 func (j *job) GetName() string {
@@ -214,12 +202,15 @@ func (j *job) DoBackup(logCh chan logger.LogRecord, tmpDir string) error {
 	var errs *multierror.Error
 
 	for ofsPart, tgt := range j.targets {
+		startTime := time.Now()
+
 		j.SetOfsMetrics(ofsPart, map[string]float64{
-			metrics.BackupOk:     float64(0),
-			metrics.BackupTime:   float64(0),
-			metrics.DeliveryOk:   float64(0),
-			metrics.DeliveryTime: float64(0),
-			metrics.BackupSize:   float64(0),
+			metrics.BackupOk:        float64(0),
+			metrics.BackupTime:      float64(0),
+			metrics.DeliveryOk:      float64(0),
+			metrics.DeliveryTime:    float64(0),
+			metrics.BackupSize:      float64(0),
+			metrics.BackupTimestamp: float64(startTime.Unix()),
 		})
 
 		tmpBackupFile := misc.GetFileFullPath(tmpDir, ofsPart, "tar", "", tgt.gzip)
@@ -230,7 +221,6 @@ func (j *job) DoBackup(logCh chan logger.LogRecord, tmpDir string) error {
 			continue
 		}
 
-		startTime := time.Now()
 		if err = targz.Tar(targz.TarOpts{
 			Src:         tgt.path,
 			Dst:         tmpBackupFile,

--- a/modules/backup/external/external.go
+++ b/modules/backup/external/external.go
@@ -25,7 +25,6 @@ type job struct {
 	storages         interfaces.Storages
 	dumpedObjects    map[string]interfaces.DumpObject
 	appMetrics       *metrics.Data
-	jobMetrics       metrics.JobData
 }
 
 type JobParams struct {
@@ -38,7 +37,6 @@ type JobParams struct {
 	SkipBackupRotate bool
 	Storages         interfaces.Storages
 	Metrics          *metrics.Data
-	OldMetrics       *metrics.Data
 }
 
 func Init(jp JobParams) (interfaces.Job, error) {
@@ -53,37 +51,28 @@ func Init(jp JobParams) (interfaces.Job, error) {
 		skipBackupRotate: jp.SkipBackupRotate,
 		storages:         jp.Storages,
 		dumpedObjects:    make(map[string]interfaces.DumpObject),
-		appMetrics:       jp.Metrics,
-		jobMetrics: metrics.JobData{
-			JobName:       jp.Name,
-			JobType:       misc.External,
-			TargetMetrics: make(map[string]metrics.TargetData),
-		},
+		appMetrics: jp.Metrics.RegisterJob(
+			metrics.JobData{
+				JobName:       jp.Name,
+				JobType:       misc.External,
+				TargetMetrics: make(map[string]metrics.TargetData),
+			},
+		),
 	}
 
-	ojm := jp.OldMetrics.GetMetrics(jp.Name)
-	if otm, ok := ojm.TargetMetrics[jp.Name]; ok {
-		j.jobMetrics.TargetMetrics[jp.Name] = otm
-	} else {
-		j.jobMetrics.TargetMetrics[jp.Name] = metrics.TargetData{
-			Source: "",
-			Target: "",
-			Values: make(map[string]float64),
-		}
+	j.appMetrics.Job[j.name].TargetMetrics[jp.Name] = metrics.TargetData{
+		Source: "",
+		Target: "",
+		Values: make(map[string]float64),
 	}
 
-	j.ExportMetrics()
 	return &j, nil
 }
 
 func (j *job) SetOfsMetrics(_ string, metrics map[string]float64) {
 	for m, v := range metrics {
-		j.jobMetrics.TargetMetrics[j.name].Values[m] = v
+		j.appMetrics.Job[j.name].TargetMetrics[j.name].Values[m] = v
 	}
-}
-
-func (j *job) ExportMetrics() {
-	j.appMetrics.JobMetricsSet(j.jobMetrics)
 }
 
 func (j *job) GetName() string {
@@ -145,12 +134,15 @@ func (j *job) DoBackup(logCh chan logger.LogRecord, _ string) (err error) {
 
 	var stderr, stdout bytes.Buffer
 
+	startTime := time.Now()
+
 	j.SetOfsMetrics("", map[string]float64{
-		metrics.BackupOk:     float64(0),
-		metrics.BackupTime:   float64(0),
-		metrics.DeliveryOk:   float64(0),
-		metrics.DeliveryTime: float64(0),
-		metrics.BackupSize:   float64(0),
+		metrics.BackupOk:        float64(0),
+		metrics.BackupTime:      float64(0),
+		metrics.DeliveryOk:      float64(0),
+		metrics.DeliveryTime:    float64(0),
+		metrics.BackupSize:      float64(0),
+		metrics.BackupTimestamp: float64(startTime.Unix()),
 	})
 
 	defer func() {
@@ -174,7 +166,6 @@ func (j *job) DoBackup(logCh chan logger.LogRecord, _ string) (err error) {
 	logCh <- logger.Log(j.name, "").Debugf("Dump cmd: %s", cmd.String())
 
 	logCh <- logger.Log(j.name, "").Infof("Starting of `%s`", j.dumpCmd)
-	startTime := time.Now()
 	if err = cmd.Run(); err != nil {
 		j.SetOfsMetrics("", map[string]float64{
 			metrics.BackupTime: float64(time.Since(startTime).Nanoseconds() / 1e6),

--- a/modules/backup/inc_files/inc_files.go
+++ b/modules/backup/inc_files/inc_files.go
@@ -32,7 +32,6 @@ type job struct {
 	targets         map[string]target
 	dumpedObjects   map[string]interfaces.DumpObject
 	appMetrics      *metrics.Data
-	jobMetrics      metrics.JobData
 }
 
 type target struct {
@@ -51,7 +50,6 @@ type JobParams struct {
 	Storages        interfaces.Storages
 	Sources         []SourceParams
 	Metrics         *metrics.Data
-	OldMetrics      *metrics.Data
 }
 
 type SourceParams struct {
@@ -77,15 +75,14 @@ func Init(jp JobParams) (interfaces.Job, error) {
 		storages:        jp.Storages,
 		dumpedObjects:   make(map[string]interfaces.DumpObject),
 		targets:         make(map[string]target),
-		appMetrics:      jp.Metrics,
-		jobMetrics: metrics.JobData{
-			JobName:       jp.Name,
-			JobType:       misc.IncFiles,
-			TargetMetrics: make(map[string]metrics.TargetData),
-		},
+		appMetrics: jp.Metrics.RegisterJob(
+			metrics.JobData{
+				JobName:       jp.Name,
+				JobType:       misc.IncFiles,
+				TargetMetrics: make(map[string]metrics.TargetData),
+			},
+		),
 	}
-
-	ojm := jp.OldMetrics.GetMetrics(jp.Name)
 
 	for _, src := range jp.Sources {
 
@@ -126,32 +123,23 @@ func Init(jp JobParams) (interfaces.Job, error) {
 						saveAbsPath: src.SaveAbsPath,
 						excludes:    excludes,
 					}
-					if otm, ok := ojm.TargetMetrics[ofs]; ok {
-						j.jobMetrics.TargetMetrics[ofs] = otm
-					} else {
-						j.jobMetrics.TargetMetrics[ofs] = metrics.TargetData{
-							Source: src.Name,
-							Target: ofsPart,
-							Values: make(map[string]float64),
-						}
+					j.appMetrics.Job[jp.Name].TargetMetrics[ofs] = metrics.TargetData{
+						Source: src.Name,
+						Target: ofsPart,
+						Values: make(map[string]float64),
 					}
 				}
 			}
 		}
 	}
 
-	j.ExportMetrics()
 	return &j, nil
 }
 
 func (j *job) SetOfsMetrics(ofs string, metricsMap map[string]float64) {
 	for m, v := range metricsMap {
-		j.jobMetrics.TargetMetrics[ofs].Values[m] = v
+		j.appMetrics.Job[j.name].TargetMetrics[ofs].Values[m] = v
 	}
-}
-
-func (j *job) ExportMetrics() {
-	j.appMetrics.JobMetricsSet(j.jobMetrics)
 }
 
 func (j *job) GetName() string {
@@ -212,12 +200,15 @@ func (j *job) DoBackup(logCh chan logger.LogRecord, tmpDir string) error {
 	var errs *multierror.Error
 
 	for ofsPart, tgt := range j.targets {
+		startTime := time.Now()
+
 		j.SetOfsMetrics(ofsPart, map[string]float64{
-			metrics.BackupOk:     float64(0),
-			metrics.BackupTime:   float64(0),
-			metrics.DeliveryOk:   float64(0),
-			metrics.DeliveryTime: float64(0),
-			metrics.BackupSize:   float64(0),
+			metrics.BackupOk:        float64(0),
+			metrics.BackupTime:      float64(0),
+			metrics.DeliveryOk:      float64(0),
+			metrics.DeliveryTime:    float64(0),
+			metrics.BackupSize:      float64(0),
+			metrics.BackupTimestamp: float64(startTime.Unix()),
 		})
 
 		tmpBackupFile := misc.GetFileFullPath(tmpDir, ofsPart, "tar", "", tgt.gzip)
@@ -245,7 +236,6 @@ func (j *job) DoBackup(logCh chan logger.LogRecord, tmpDir string) error {
 			}
 		}
 
-		startTime := time.Now()
 		if err = targz.Tar(targz.TarOpts{
 			Src:         tgt.path,
 			Dst:         tmpBackupFile,

--- a/modules/backup/mongodump/mongodump.go
+++ b/modules/backup/mongodump/mongodump.go
@@ -34,7 +34,6 @@ type job struct {
 	targets          map[string]target
 	dumpedObjects    map[string]interfaces.DumpObject
 	appMetrics       *metrics.Data
-	jobMetrics       metrics.JobData
 }
 
 type target struct {
@@ -56,7 +55,6 @@ type JobParams struct {
 	Storages         interfaces.Storages
 	Sources          []SourceParams
 	Metrics          *metrics.Data
-	OldMetrics       *metrics.Data
 }
 
 type SourceParams struct {
@@ -91,15 +89,14 @@ func Init(jp JobParams) (interfaces.Job, error) {
 		storages:         jp.Storages,
 		targets:          make(map[string]target),
 		dumpedObjects:    make(map[string]interfaces.DumpObject),
-		appMetrics:       jp.Metrics,
-		jobMetrics: metrics.JobData{
-			JobName:       jp.Name,
-			JobType:       misc.MongoDB,
-			TargetMetrics: make(map[string]metrics.TargetData),
-		},
+		appMetrics: jp.Metrics.RegisterJob(
+			metrics.JobData{
+				JobName:       jp.Name,
+				JobType:       misc.MongoDB,
+				TargetMetrics: make(map[string]metrics.TargetData),
+			},
+		),
 	}
-
-	ojm := jp.OldMetrics.GetMetrics(jp.Name)
 
 	for _, src := range jp.Sources {
 
@@ -163,30 +160,21 @@ func Init(jp JobParams) (interfaces.Job, error) {
 				gzip:        src.Gzip,
 				connOpts:    src.ConnectParams,
 			}
-			if otm, ok := ojm.TargetMetrics[ofs]; ok {
-				j.jobMetrics.TargetMetrics[ofs] = otm
-			} else {
-				j.jobMetrics.TargetMetrics[ofs] = metrics.TargetData{
-					Source: src.Name,
-					Target: db,
-					Values: make(map[string]float64),
-				}
+			j.appMetrics.Job[j.name].TargetMetrics[ofs] = metrics.TargetData{
+				Source: src.Name,
+				Target: db,
+				Values: make(map[string]float64),
 			}
 		}
 	}
 
-	j.ExportMetrics()
 	return &j, nil
 }
 
 func (j *job) SetOfsMetrics(ofs string, metricsMap map[string]float64) {
 	for m, v := range metricsMap {
-		j.jobMetrics.TargetMetrics[ofs].Values[m] = v
+		j.appMetrics.Job[j.name].TargetMetrics[ofs].Values[m] = v
 	}
-}
-
-func (j *job) ExportMetrics() {
-	j.appMetrics.JobMetricsSet(j.jobMetrics)
 }
 
 func (j *job) GetName() string {
@@ -247,12 +235,15 @@ func (j *job) DoBackup(logCh chan logger.LogRecord, tmpDir string) error {
 	var errs *multierror.Error
 
 	for ofsPart, tgt := range j.targets {
+		startTime := time.Now()
+
 		j.SetOfsMetrics(ofsPart, map[string]float64{
-			metrics.BackupOk:     float64(0),
-			metrics.BackupTime:   float64(0),
-			metrics.DeliveryOk:   float64(0),
-			metrics.DeliveryTime: float64(0),
-			metrics.BackupSize:   float64(0),
+			metrics.BackupOk:        float64(0),
+			metrics.BackupTime:      float64(0),
+			metrics.DeliveryOk:      float64(0),
+			metrics.DeliveryTime:    float64(0),
+			metrics.BackupSize:      float64(0),
+			metrics.BackupTimestamp: float64(startTime.Unix()),
 		})
 
 		tmpBackupFile := misc.GetFileFullPath(tmpDir, ofsPart, "tar", "", tgt.gzip)
@@ -263,7 +254,6 @@ func (j *job) DoBackup(logCh chan logger.LogRecord, tmpDir string) error {
 			continue
 		}
 
-		startTime := time.Now()
 		if err := j.createTmpBackup(logCh, tmpBackupFile, tgt); err != nil {
 			j.SetOfsMetrics(ofsPart, map[string]float64{
 				metrics.BackupTime: float64(time.Since(startTime).Nanoseconds() / 1e6),

--- a/modules/backup/mysql/mysql.go
+++ b/modules/backup/mysql/mysql.go
@@ -35,7 +35,6 @@ type job struct {
 	dumpedObjects    map[string]interfaces.DumpObject
 	authFilesKeys    map[string][]byte
 	appMetrics       *metrics.Data
-	jobMetrics       metrics.JobData
 }
 
 type target struct {
@@ -58,7 +57,6 @@ type JobParams struct {
 	Storages         interfaces.Storages
 	Sources          []SourceParams
 	Metrics          *metrics.Data
-	OldMetrics       *metrics.Data
 }
 
 type SourceParams struct {
@@ -88,15 +86,14 @@ func Init(jp JobParams) (interfaces.Job, error) {
 		storages:         jp.Storages,
 		targets:          make(map[string]target),
 		dumpedObjects:    make(map[string]interfaces.DumpObject),
-		appMetrics:       jp.Metrics,
-		jobMetrics: metrics.JobData{
-			JobName:       jp.Name,
-			JobType:       misc.Mysql,
-			TargetMetrics: make(map[string]metrics.TargetData),
-		},
+		appMetrics: jp.Metrics.RegisterJob(
+			metrics.JobData{
+				JobName:       jp.Name,
+				JobType:       misc.Mysql,
+				TargetMetrics: make(map[string]metrics.TargetData),
+			},
+		),
 	}
-
-	ojm := jp.OldMetrics.GetMetrics(jp.Name)
 
 	for _, src := range jp.Sources {
 
@@ -138,30 +135,21 @@ func Init(jp JobParams) (interfaces.Job, error) {
 				gzip:         src.Gzip,
 				isSlave:      src.IsSlave,
 			}
-			if otm, ok := ojm.TargetMetrics[ofs]; ok {
-				j.jobMetrics.TargetMetrics[ofs] = otm
-			} else {
-				j.jobMetrics.TargetMetrics[ofs] = metrics.TargetData{
-					Source: src.Name,
-					Target: db,
-					Values: make(map[string]float64),
-				}
+			j.appMetrics.Job[j.name].TargetMetrics[ofs] = metrics.TargetData{
+				Source: src.Name,
+				Target: db,
+				Values: make(map[string]float64),
 			}
 		}
 	}
 
-	j.ExportMetrics()
 	return &j, nil
 }
 
 func (j *job) SetOfsMetrics(ofs string, metricsMap map[string]float64) {
 	for m, v := range metricsMap {
-		j.jobMetrics.TargetMetrics[ofs].Values[m] = v
+		j.appMetrics.Job[j.name].TargetMetrics[ofs].Values[m] = v
 	}
-}
-
-func (j *job) ExportMetrics() {
-	j.appMetrics.JobMetricsSet(j.jobMetrics)
 }
 
 func (j *job) GetName() string {
@@ -222,12 +210,15 @@ func (j *job) DoBackup(logCh chan logger.LogRecord, tmpDir string) error {
 	var errs *multierror.Error
 
 	for ofsPart, tgt := range j.targets {
+		startTime := time.Now()
+
 		j.SetOfsMetrics(ofsPart, map[string]float64{
-			metrics.BackupOk:     float64(0),
-			metrics.BackupTime:   float64(0),
-			metrics.DeliveryOk:   float64(0),
-			metrics.DeliveryTime: float64(0),
-			metrics.BackupSize:   float64(0),
+			metrics.BackupOk:        float64(0),
+			metrics.BackupTime:      float64(0),
+			metrics.DeliveryOk:      float64(0),
+			metrics.DeliveryTime:    float64(0),
+			metrics.BackupSize:      float64(0),
+			metrics.BackupTimestamp: float64(startTime.Unix()),
 		})
 
 		tmpBackupFile := misc.GetFileFullPath(tmpDir, ofsPart, "sql", "", tgt.gzip)
@@ -238,7 +229,6 @@ func (j *job) DoBackup(logCh chan logger.LogRecord, tmpDir string) error {
 			continue
 		}
 
-		startTime := time.Now()
 		if err = j.createTmpBackup(logCh, tmpBackupFile, tgt); err != nil {
 			j.SetOfsMetrics(ofsPart, map[string]float64{
 				metrics.BackupTime: float64(time.Since(startTime).Nanoseconds() / 1e6),

--- a/modules/backup/mysql_xtrabackup/mysql_xtrabackup.go
+++ b/modules/backup/mysql_xtrabackup/mysql_xtrabackup.go
@@ -35,7 +35,6 @@ type job struct {
 	targets          map[string]target
 	dumpedObjects    map[string]interfaces.DumpObject
 	appMetrics       *metrics.Data
-	jobMetrics       metrics.JobData
 }
 
 type target struct {
@@ -57,7 +56,6 @@ type JobParams struct {
 	Storages         interfaces.Storages
 	Sources          []SourceParams
 	Metrics          *metrics.Data
-	OldMetrics       *metrics.Data
 }
 
 type SourceParams struct {
@@ -92,15 +90,14 @@ func Init(jp JobParams) (interfaces.Job, error) {
 		storages:         jp.Storages,
 		targets:          make(map[string]target),
 		dumpedObjects:    make(map[string]interfaces.DumpObject),
-		appMetrics:       jp.Metrics,
-		jobMetrics: metrics.JobData{
-			JobName:       jp.Name,
-			JobType:       misc.MysqlXtrabackup,
-			TargetMetrics: make(map[string]metrics.TargetData),
-		},
+		appMetrics: jp.Metrics.RegisterJob(
+			metrics.JobData{
+				JobName:       jp.Name,
+				JobType:       misc.MysqlXtrabackup,
+				TargetMetrics: make(map[string]metrics.TargetData),
+			},
+		),
 	}
-
-	ojm := jp.OldMetrics.GetMetrics(jp.Name)
 
 	for _, src := range jp.Sources {
 
@@ -126,29 +123,20 @@ func Init(jp JobParams) (interfaces.Job, error) {
 			isSlave:         src.IsSlave,
 			prepare:         src.Prepare,
 		}
-		if otm, ok := ojm.TargetMetrics[src.Name]; ok {
-			j.jobMetrics.TargetMetrics[src.Name] = otm
-		} else {
-			j.jobMetrics.TargetMetrics[src.Name] = metrics.TargetData{
-				Source: src.Name,
-				Target: "",
-				Values: make(map[string]float64),
-			}
+		j.appMetrics.Job[j.name].TargetMetrics[src.Name] = metrics.TargetData{
+			Source: src.Name,
+			Target: "",
+			Values: make(map[string]float64),
 		}
 	}
 
-	j.ExportMetrics()
 	return &j, nil
 }
 
 func (j *job) SetOfsMetrics(ofs string, metricsMap map[string]float64) {
 	for m, v := range metricsMap {
-		j.jobMetrics.TargetMetrics[ofs].Values[m] = v
+		j.appMetrics.Job[j.name].TargetMetrics[ofs].Values[m] = v
 	}
-}
-
-func (j *job) ExportMetrics() {
-	j.appMetrics.JobMetricsSet(j.jobMetrics)
 }
 
 func (j *job) GetName() string {
@@ -209,12 +197,15 @@ func (j *job) DoBackup(logCh chan logger.LogRecord, tmpDir string) error {
 	var errs *multierror.Error
 
 	for ofsPart, tgt := range j.targets {
+		startTime := time.Now()
+
 		j.SetOfsMetrics(ofsPart, map[string]float64{
-			metrics.BackupOk:     float64(0),
-			metrics.BackupTime:   float64(0),
-			metrics.DeliveryOk:   float64(0),
-			metrics.DeliveryTime: float64(0),
-			metrics.BackupSize:   float64(0),
+			metrics.BackupOk:        float64(0),
+			metrics.BackupTime:      float64(0),
+			metrics.DeliveryOk:      float64(0),
+			metrics.DeliveryTime:    float64(0),
+			metrics.BackupSize:      float64(0),
+			metrics.BackupTimestamp: float64(startTime.Unix()),
 		})
 
 		tmpBackupFile := misc.GetFileFullPath(tmpDir, ofsPart, "tar", "", tgt.gzip)
@@ -225,7 +216,6 @@ func (j *job) DoBackup(logCh chan logger.LogRecord, tmpDir string) error {
 			continue
 		}
 
-		startTime := time.Now()
 		if err = j.createTmpBackup(logCh, tmpBackupFile, ofsPart, tgt); err != nil {
 			j.SetOfsMetrics(ofsPart, map[string]float64{
 				metrics.BackupTime: float64(time.Since(startTime).Nanoseconds() / 1e6),

--- a/modules/backup/psql/psql.go
+++ b/modules/backup/psql/psql.go
@@ -34,7 +34,6 @@ type job struct {
 	targets          map[string]target
 	dumpedObjects    map[string]interfaces.DumpObject
 	appMetrics       *metrics.Data
-	jobMetrics       metrics.JobData
 }
 
 type target struct {
@@ -55,7 +54,6 @@ type JobParams struct {
 	Storages         interfaces.Storages
 	Sources          []SourceParams
 	Metrics          *metrics.Data
-	OldMetrics       *metrics.Data
 }
 
 type SourceParams struct {
@@ -86,15 +84,14 @@ func Init(jp JobParams) (interfaces.Job, error) {
 		storages:         jp.Storages,
 		targets:          make(map[string]target),
 		dumpedObjects:    make(map[string]interfaces.DumpObject),
-		appMetrics:       jp.Metrics,
-		jobMetrics: metrics.JobData{
-			JobName:       jp.Name,
-			JobType:       misc.Postgresql,
-			TargetMetrics: make(map[string]metrics.TargetData),
-		},
+		appMetrics: jp.Metrics.RegisterJob(
+			metrics.JobData{
+				JobName:       jp.Name,
+				JobType:       misc.Postgresql,
+				TargetMetrics: make(map[string]metrics.TargetData),
+			},
+		),
 	}
-
-	ojm := jp.OldMetrics.GetMetrics(jp.Name)
 
 	for _, src := range jp.Sources {
 
@@ -164,30 +161,21 @@ func Init(jp JobParams) (interfaces.Job, error) {
 				extraKeys:    src.ExtraKeys,
 				gzip:         src.Gzip,
 			}
-			if otm, ok := ojm.TargetMetrics[ofs]; ok {
-				j.jobMetrics.TargetMetrics[ofs] = otm
-			} else {
-				j.jobMetrics.TargetMetrics[ofs] = metrics.TargetData{
-					Source: src.Name,
-					Target: db,
-					Values: make(map[string]float64),
-				}
+			j.appMetrics.Job[j.name].TargetMetrics[ofs] = metrics.TargetData{
+				Source: src.Name,
+				Target: db,
+				Values: make(map[string]float64),
 			}
 		}
 	}
 
-	j.ExportMetrics()
 	return &j, nil
 }
 
 func (j *job) SetOfsMetrics(ofs string, metricsMap map[string]float64) {
 	for m, v := range metricsMap {
-		j.jobMetrics.TargetMetrics[ofs].Values[m] = v
+		j.appMetrics.Job[j.name].TargetMetrics[ofs].Values[m] = v
 	}
-}
-
-func (j *job) ExportMetrics() {
-	j.appMetrics.JobMetricsSet(j.jobMetrics)
 }
 
 func (j *job) GetName() string {
@@ -248,12 +236,15 @@ func (j *job) DoBackup(logCh chan logger.LogRecord, tmpDir string) error {
 	var errs *multierror.Error
 
 	for ofsPart, tgt := range j.targets {
+		startTime := time.Now()
+
 		j.SetOfsMetrics(ofsPart, map[string]float64{
-			metrics.BackupOk:     float64(0),
-			metrics.BackupTime:   float64(0),
-			metrics.DeliveryOk:   float64(0),
-			metrics.DeliveryTime: float64(0),
-			metrics.BackupSize:   float64(0),
+			metrics.BackupOk:        float64(0),
+			metrics.BackupTime:      float64(0),
+			metrics.DeliveryOk:      float64(0),
+			metrics.DeliveryTime:    float64(0),
+			metrics.BackupSize:      float64(0),
+			metrics.BackupTimestamp: float64(startTime.Unix()),
 		})
 
 		tmpBackupFile := misc.GetFileFullPath(tmpDir, ofsPart, "sql", "", tgt.gzip)
@@ -264,7 +255,6 @@ func (j *job) DoBackup(logCh chan logger.LogRecord, tmpDir string) error {
 			continue
 		}
 
-		startTime := time.Now()
 		if err = j.createTmpBackup(logCh, tmpBackupFile, tgt); err != nil {
 			j.SetOfsMetrics(ofsPart, map[string]float64{
 				metrics.BackupTime: float64(time.Since(startTime).Nanoseconds() / 1e6),

--- a/modules/backup/psql_basebackup/psql_basebackup.go
+++ b/modules/backup/psql_basebackup/psql_basebackup.go
@@ -33,7 +33,6 @@ type job struct {
 	targets          map[string]target
 	dumpedObjects    map[string]interfaces.DumpObject
 	appMetrics       *metrics.Data
-	jobMetrics       metrics.JobData
 }
 
 type target struct {
@@ -52,7 +51,6 @@ type JobParams struct {
 	Storages         interfaces.Storages
 	Sources          []SourceParams
 	Metrics          *metrics.Data
-	OldMetrics       *metrics.Data
 }
 
 type SourceParams struct {
@@ -84,15 +82,14 @@ func Init(jp JobParams) (interfaces.Job, error) {
 		storages:         jp.Storages,
 		targets:          make(map[string]target),
 		dumpedObjects:    make(map[string]interfaces.DumpObject),
-		appMetrics:       jp.Metrics,
+		appMetrics: jp.Metrics.RegisterJob(
+			metrics.JobData{
+				JobName:       jp.Name,
+				JobType:       misc.PostgresqlBasebackup,
+				TargetMetrics: make(map[string]metrics.TargetData),
+			},
+		),
 	}
-
-	j.jobMetrics = metrics.JobData{
-		JobName:       jp.Name,
-		JobType:       j.GetType(),
-		TargetMetrics: make(map[string]metrics.TargetData),
-	}
-	ojm := jp.OldMetrics.GetMetrics(jp.Name)
 
 	for _, src := range jp.Sources {
 
@@ -124,29 +121,20 @@ func Init(jp JobParams) (interfaces.Job, error) {
 			gzip:      src.Gzip,
 			connUrl:   connUrl,
 		}
-		if otm, ok := ojm.TargetMetrics[src.Name]; ok {
-			j.jobMetrics.TargetMetrics[src.Name] = otm
-		} else {
-			j.jobMetrics.TargetMetrics[src.Name] = metrics.TargetData{
-				Source: src.Name,
-				Target: "",
-				Values: make(map[string]float64),
-			}
+		j.appMetrics.Job[j.name].TargetMetrics[src.Name] = metrics.TargetData{
+			Source: src.Name,
+			Target: "",
+			Values: make(map[string]float64),
 		}
 	}
 
-	j.ExportMetrics()
 	return &j, nil
 }
 
 func (j *job) SetOfsMetrics(ofs string, metricsMap map[string]float64) {
 	for m, v := range metricsMap {
-		j.jobMetrics.TargetMetrics[ofs].Values[m] = v
+		j.appMetrics.Job[j.name].TargetMetrics[ofs].Values[m] = v
 	}
-}
-
-func (j *job) ExportMetrics() {
-	j.appMetrics.JobMetricsSet(j.jobMetrics)
 }
 
 func (j *job) GetName() string {
@@ -207,12 +195,15 @@ func (j *job) DoBackup(logCh chan logger.LogRecord, tmpDir string) error {
 	var errs *multierror.Error
 
 	for ofsPart, tgt := range j.targets {
+		startTime := time.Now()
+
 		j.SetOfsMetrics(ofsPart, map[string]float64{
-			metrics.BackupOk:     float64(0),
-			metrics.BackupTime:   float64(0),
-			metrics.DeliveryOk:   float64(0),
-			metrics.DeliveryTime: float64(0),
-			metrics.BackupSize:   float64(0),
+			metrics.BackupOk:        float64(0),
+			metrics.BackupTime:      float64(0),
+			metrics.DeliveryOk:      float64(0),
+			metrics.DeliveryTime:    float64(0),
+			metrics.BackupSize:      float64(0),
+			metrics.BackupTimestamp: float64(startTime.Unix()),
 		})
 
 		tmpBackupFile := misc.GetFileFullPath(tmpDir, ofsPart, "tar", "", tgt.gzip)
@@ -223,7 +214,6 @@ func (j *job) DoBackup(logCh chan logger.LogRecord, tmpDir string) error {
 			continue
 		}
 
-		startTime := time.Now()
 		if err = j.createTmpBackup(logCh, tmpBackupFile, ofsPart, tgt); err != nil {
 			j.SetOfsMetrics(ofsPart, map[string]float64{
 				metrics.BackupTime: float64(time.Since(startTime).Nanoseconds() / 1e6),

--- a/modules/backup/redis/redis.go
+++ b/modules/backup/redis/redis.go
@@ -28,7 +28,6 @@ type job struct {
 	deferredCopying  bool
 	diskRateLimit    int64
 	appMetrics       *metrics.Data
-	jobMetrics       metrics.JobData
 	storages         interfaces.Storages
 	targets          map[string]target
 	dumpedObjects    map[string]interfaces.DumpObject
@@ -49,7 +48,6 @@ type JobParams struct {
 	Storages         interfaces.Storages
 	Sources          []SourceParams
 	Metrics          *metrics.Data
-	OldMetrics       *metrics.Data
 }
 
 type SourceParams struct {
@@ -76,15 +74,14 @@ func Init(jp JobParams) (interfaces.Job, error) {
 		storages:         jp.Storages,
 		targets:          make(map[string]target),
 		dumpedObjects:    make(map[string]interfaces.DumpObject),
-		appMetrics:       jp.Metrics,
-		jobMetrics: metrics.JobData{
-			JobName:       jp.Name,
-			JobType:       misc.Redis,
-			TargetMetrics: make(map[string]metrics.TargetData),
-		},
+		appMetrics: jp.Metrics.RegisterJob(
+			metrics.JobData{
+				JobName:       jp.Name,
+				JobType:       misc.Redis,
+				TargetMetrics: make(map[string]metrics.TargetData),
+			},
+		),
 	}
-
-	ojm := jp.OldMetrics.GetMetrics(jp.Name)
 
 	for _, src := range jp.Sources {
 
@@ -98,29 +95,20 @@ func Init(jp JobParams) (interfaces.Job, error) {
 			gzip: src.Gzip,
 			dsn:  dsn,
 		}
-		if otm, ok := ojm.TargetMetrics[src.Name]; ok {
-			j.jobMetrics.TargetMetrics[src.Name] = otm
-		} else {
-			j.jobMetrics.TargetMetrics[src.Name] = metrics.TargetData{
-				Source: src.Name,
-				Target: "",
-				Values: make(map[string]float64),
-			}
+		j.appMetrics.Job[j.name].TargetMetrics[src.Name] = metrics.TargetData{
+			Source: src.Name,
+			Target: "",
+			Values: make(map[string]float64),
 		}
 	}
 
-	j.ExportMetrics()
 	return &j, nil
 }
 
 func (j *job) SetOfsMetrics(ofs string, metricsMap map[string]float64) {
 	for m, v := range metricsMap {
-		j.jobMetrics.TargetMetrics[ofs].Values[m] = v
+		j.appMetrics.Job[j.name].TargetMetrics[ofs].Values[m] = v
 	}
-}
-
-func (j *job) ExportMetrics() {
-	j.appMetrics.JobMetricsSet(j.jobMetrics)
 }
 
 func (j *job) GetName() string {
@@ -181,12 +169,15 @@ func (j *job) DoBackup(logCh chan logger.LogRecord, tmpDir string) error {
 	var errs *multierror.Error
 
 	for ofsPart, tgt := range j.targets {
+		startTime := time.Now()
+
 		j.SetOfsMetrics(ofsPart, map[string]float64{
-			metrics.BackupOk:     float64(0),
-			metrics.BackupTime:   float64(0),
-			metrics.DeliveryOk:   float64(0),
-			metrics.DeliveryTime: float64(0),
-			metrics.BackupSize:   float64(0),
+			metrics.BackupOk:        float64(0),
+			metrics.BackupTime:      float64(0),
+			metrics.DeliveryOk:      float64(0),
+			metrics.DeliveryTime:    float64(0),
+			metrics.BackupSize:      float64(0),
+			metrics.BackupTimestamp: float64(startTime.Unix()),
 		})
 
 		tmpBackupFile := misc.GetFileFullPath(tmpDir, ofsPart, "rdb", "", tgt.gzip)
@@ -197,7 +188,6 @@ func (j *job) DoBackup(logCh chan logger.LogRecord, tmpDir string) error {
 			continue
 		}
 
-		startTime := time.Now()
 		if err = j.createTmpBackup(logCh, tmpBackupFile, ofsPart, tgt); err != nil {
 			j.SetOfsMetrics(ofsPart, map[string]float64{
 				metrics.BackupTime: float64(time.Since(startTime).Nanoseconds() / 1e6),

--- a/modules/cmd_handler/start_backup/start_backup.go
+++ b/modules/cmd_handler/start_backup/start_backup.go
@@ -57,7 +57,17 @@ func Init(o Opts) *startBackup {
 }
 
 func (sb *startBackup) Run() {
-	var errs *multierror.Error
+	var (
+		err  error
+		errs *multierror.Error
+	)
+
+	defer func() {
+		if err := sb.metricsData.SaveFile(); err != nil {
+			sb.evCh <- logger.Log("", "").Errorf("Failed to save metrics to file: %v", err)
+		}
+		sb.done <- err
+	}()
 
 	if sb.initErr != nil {
 		sb.evCh <- logger.Log("", "").Errorf("Backup plan initialised with errors: %v", sb.initErr)
@@ -83,7 +93,6 @@ func (sb *startBackup) Run() {
 	if err != nil {
 		err = fmt.Errorf("Can't start nxs-backup. Another nxs-backup process already running. ")
 		sb.evCh <- logger.Log("", "").Error(err)
-		sb.done <- err
 		return
 	}
 	defer func() { _ = lock.Unlock() }()
@@ -132,13 +141,8 @@ func (sb *startBackup) Run() {
 	}
 
 	sb.evCh <- logger.Log("", "").Infof("Backup finished.\n")
-	if sb.metricsData != nil {
-		if err = sb.metricsData.SaveFile(); err != nil {
-			sb.evCh <- logger.Log("", "").Errorf("Failed to save metrics to file: %v", err)
-		}
-	}
+
 	if errs.ErrorOrNil() != nil {
-		sb.done <- fmt.Errorf("Some of backups failed with next errors:\n%w", errs)
+		err = fmt.Errorf("Some of backups failed with next errors:\n%w", errs)
 	}
-	sb.done <- nil
 }

--- a/modules/metrics/exporter.go
+++ b/modules/metrics/exporter.go
@@ -37,6 +37,11 @@ func InitExporter(s ExporterOpts) *Exporter {
 			"Backup collection time",
 			[]string{"project", "server", "job_name", "job_type", "source", "target"}, nil,
 		),
+		BackupTimestamp: prometheus.NewDesc(
+			prometheus.BuildFQName("nxs_backup", "creation", "ts"),
+			"Backup creation timestamp",
+			[]string{"project", "server", "job_name", "job_type", "source", "target"}, nil,
+		),
 		DeliveryOk: prometheus.NewDesc(
 			prometheus.BuildFQName("nxs_backup", "delivery", "success"),
 			"Backup delivery finished successfully",

--- a/modules/metrics/metrics.go
+++ b/modules/metrics/metrics.go
@@ -16,6 +16,7 @@ const (
 
 	BackupOk        = "backup_ok"
 	BackupTime      = "backup_time"
+	BackupTimestamp = "backup_timestamp"
 	BackupSize      = "size"
 	DeliveryOk      = "delivery_ok"
 	DeliveryTime    = "delivery_time"
@@ -26,6 +27,7 @@ type Data struct {
 	Project             string
 	Server              string
 	NewVersionAvailable float64
+	Enabled             bool
 
 	Job map[string]JobData
 
@@ -49,48 +51,49 @@ type DataOpts struct {
 	Server              string
 	MetricsFile         string
 	NewVersionAvailable float64
+	Enabled             bool
 }
 
-// InitData initializes the metrics data by reading from a file and
+// InitData initializes the metrics data by
 // creating a new Data instance with the provided options.
-// It returns the old metrics data, the new initialized data, and any error encountered.
-//
-// TODO Remove the reading of old metrics for the save step.
-func InitData(opts DataOpts) (data, oldMetrics *Data, err error) {
-	oldMetrics, err = readFile(opts.MetricsFile)
-	if err != nil {
-		return
-	}
-
-	data = &Data{
+func InitData(opts DataOpts) *Data {
+	return &Data{
 		Project:             opts.Project,
 		Server:              opts.Server,
 		NewVersionAvailable: opts.NewVersionAvailable,
+		Enabled:             opts.Enabled,
 		Job:                 make(map[string]JobData),
 		metricsFile:         opts.MetricsFile,
 	}
-
-	return
 }
 
 func (md *Data) MetricFilePath() string {
 	return md.metricsFile
 }
 
-func (md *Data) GetMetrics(jobName string) JobData {
-	if job, ok := md.Job[jobName]; ok {
-		return job
-	}
-	return JobData{
-		TargetMetrics: make(map[string]TargetData),
-	}
-}
-
-func (md *Data) JobMetricsSet(jd JobData) {
+func (md *Data) RegisterJob(jd JobData) *Data {
 	md.Job[jd.JobName] = jd
+	return md
 }
 
 func (md *Data) SaveFile() error {
+	//skip if metrics disabled
+	if !md.Enabled {
+		return nil
+	}
+
+	od, err := readFile(md.metricsFile)
+	if err != nil {
+		return err
+	}
+
+	// reuse old metrics for not run jobs
+	for jobName, job := range od.Job {
+		if _, ok := md.Job[jobName]; !ok {
+			md.Job[jobName] = job
+		}
+	}
+
 	f, err := os.OpenFile(md.metricsFile, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0600)
 	if err != nil {
 		return err

